README.md

```diff
@@ -1,4 +1,102 @@
+
+## Purpose
+
+The objective of KindedJ is to provide a common set of interfaces that enable cross-library and cross-language support for all projects using evidence-based Higher Kinded Types emulation in the JVM.
+
+A Higher Kinded Type represents a value that wraps another value. They can be reasoned about as type constructors, where arity represents the number of generic parameters.
+For example, a List can contain only one parameter and is expressed as `List<T>`. A type that can contain either a value of type A or a value of type B has arity 2, and requires a type constructor with the same number of parameters, expressed as `Either<A, B>`. The group of type constructors with exactly one parameter is usually represented as `* -> *`, and every increment in arity adds a new * parameter. This way, `Either<A, B>` has a type constructor that belongs to the group `* -> * -> * `.
+
+Filling a generic parameter with a concrete value lowers the arity of the constructor by one *, so an `Either<Int, B>` belongs to `* -> *` and List<String> is `*`. All arities can be represented as arity 1 by grouping them on what's called a type function `(* -> * -> *) -> *`. This grouping is a key construct to support Higher Kinded Type emulation in the JVM.
+ 
+This representation of generic polymorphism for any arity allows library writers to create abstractions that work on any wrapper type in a completely generic fashion. For library design this means the equivalent of supporting parameters of type `F<A>`, which would take any `* -> *`.
+
+You can see some examples of generic functions using kinds in the projects supporting KindedJ listed above. You can read more about kind and type theory in [Wikipedia](https://en.wikipedia.org/wiki/Kind_(type_theory)) or [StackOverflow](https://softwareengineering.stackexchange.com/a/276861/72626).
+
+## Usage
+
+Evidence-based Higher Kinded Types (HKTs) is a way of emulating HKTs even without explicit language support. It is described in the paper [Lightweight higher-kinded Polymorphism](https://www.cl.cam.ac.uk/~jdy22/papers/lightweight-higher-kinded-polymorphism.pdf) by Jeremy Yallop and Leo White.
+
+A *datatype* is any class who we want to make into an HKT. Examples include List, Future, Optional, Stream, Observable... To consider a class a datatype it has to implement the interface `Hk<F, A>`, or provide a wrapper class that does. In this interface, the generic parameter `F` is considered the *witness* and the generic parameter `A` the *value*.
+
+A *witness* type is an empty type without constructors. This type is unique globally and exclusively used to identify a single datatype when used as a generic parameter. This type has to be different than the datatype, a datatype should not be a witness of itself. Generic parameters requiring a witness are usually represented by the characters F, G, H, or datatype specific characters like L for the left side of an Either.
```

Comment:
In eta-hkts, you can just use the original datatype as the witness by filling in the type parameters with ?. For example, `List<A>`'s witness type would be `List<?>`. It comes with a tradeoff that the number of ?'s you need grows linearly with the arity of the HKT. Other than this tradeoff are there others I'm missing?

Reply:
@rahulmutt, nope that is the tradeoff.

Reply:
You need to write different code for every arity, right? We'd like to write it once only.

Reply:
@pakoito not sure what you mean. Could you elaborate?

Reply:
@jbgi Nevermind, I derped for a second there.